**Dataset origin:** https://www.kaggle.com/devendra416/ddos-datasets \
**Feature descriptions:** https://www.unb.ca/cic/datasets/ids-2018.html \
**Dataset generation:** https://www.ijcseonline.org/pdf_paper_view.php?paper_id=4011&28-IJCSE-06600.pdf

In [1]:
!pip install dask
!pip install dask[dataframe]

  Using cached partd-1.2.0-py3-none-any.whl (19 kB)


  Using cached PyYAML-5.4.1-cp39-cp39-win_amd64.whl (213 kB)
  Using cached toolz-0.11.1-py3-none-any.whl (55 kB)
  Using cached cloudpickle-1.6.0-py3-none-any.whl (23 kB)
  Using cached locket-0.2.1-py2.py3-none-any.whl (4.1 kB)



You should consider upgrading via the 'd:\applications\ddos-dompteure\ml_vnev\scripts\python.exe -m pip install --upgrade pip' command.


  Using cached pandas-1.2.4-cp39-cp39-win_amd64.whl (9.3 MB)

You should consider upgrading via the 'd:\applications\ddos-dompteure\ml_vnev\scripts\python.exe -m pip install --upgrade pip' command.


In [2]:
import pandas as pd
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

In [4]:
import dask
import dask.dataframe as dd
import dask.array as da
df = dd.read_csv('unbalaced_20_80_dataset.csv')

In [3]:
df.head()

Unnamed: 0                                 Flow ID          Src IP  \
0     1739476   172.31.69.25-18.219.193.20-80-37882-6   18.219.193.20   
1     1822666      172.31.69.28-18.219.9.1-80-63287-6    172.31.69.28   
2      905739   172.31.69.28-52.14.136.135-80-63095-6   52.14.136.135   
3     1143064  172.31.69.28-18.216.200.189-80-52341-6  18.216.200.189   
4     1934016   172.31.69.28-18.218.55.126-80-57459-6    172.31.69.28   

   Src Port         Dst IP  Dst Port  Protocol               Timestamp  \
0     37882   172.31.69.25        80         6  16/02/2018 11:27:29 PM   
1        80     18.219.9.1     63287         6  22/02/2018 12:13:52 AM   
2     63095   172.31.69.28        80         6  22/02/2018 12:14:02 AM   
3     52341   172.31.69.28        80         6  22/02/2018 12:28:04 AM   
4        80  18.218.55.126     57459         6  22/02/2018 12:19:45 AM   

   Flow Duration  Tot Fwd Pkts  Tot Bwd Pkts  TotLen Fwd Pkts  \
0           8660             1             1              0.0   
1           5829             4             3            935.0   
2           3396             1             1              0.0   
3           2390             1             1              0.0   
4          17362             4             3            935.0   

   TotLen Bwd Pkts  Fwd Pkt Len Max  Fwd Pkt Len Min  Fwd Pkt Len Mean  \
0              0.0              0.0              0.0              0.00   
1            298.0            935.0              0.0            233.75   
2              0.0              0.0              0.0              0.00   
3              0.0              0.0              0.0              0.00   
4            314.0            935.0              0.0            233.75   

   Fwd Pkt Len Std  Bwd Pkt Len Max  Bwd Pkt Len Min  Bwd Pkt Len Mean  \
0              0.0              0.0              0.0          0.000000   
1            467.5            298.0              0.0         99.333333   
2              0.0              0.0              0.0          0.000000   
3              0.0              0.0              0.0          0.000000   
4            467.5            314.0              0.0        104.666667   

   Bwd Pkt Len Std    Flow Byts/s  Flow Pkts/s  Flow IAT Mean  Flow IAT Std  \
0         0.000000       0.000000   230.946882    8660.000000      0.000000   
1       172.050380  211528.564076  1200.892091     971.500000   2104.124973   
2         0.000000       0.000000   588.928151    3396.000000      0.000000   
3         0.000000       0.000000   836.820084    2390.000000      0.000000   
4       181.287985   71938.716738   403.179357    2893.666667   6770.577661   

   Flow IAT Max  Flow IAT Min  Fwd IAT Tot  Fwd IAT Mean  Fwd IAT Std  \
0        8660.0        8660.0          0.0      0.000000     0.000000   
1        5260.0           7.0       5822.0   1940.666667  3119.411857   
2        3396.0        3396.0          0.0      0.000000     0.000000   
3        2390.0        2390.0          0.0      0.000000     0.000000   
4       16711.0           7.0      17355.0   5785.000000  9779.680414   

   Fwd IAT Max  Fwd IAT Min  Bwd IAT Tot  Bwd IAT Mean   Bwd IAT Std  \
0          0.0          0.0          0.0           0.0      0.000000   
1       5541.0         46.0       5540.0        2770.0   3521.391770   
2          0.0          0.0          0.0           0.0      0.000000   
3          0.0          0.0          0.0           0.0      0.000000   
4      17077.0         38.0      17074.0        8537.0  11559.781659   

   Bwd IAT Max  Bwd IAT Min  Fwd PSH Flags  Bwd PSH Flags  Fwd URG Flags  \
0          0.0          0.0              0              0              0   
1       5260.0        280.0              0              0              0   
2          0.0          0.0              0              0              0   
3          0.0          0.0              0              0              0   
4      16711.0        363.0              0              0              0   

   Bwd URG Flags  Fwd Header Len  

In [3]:
df.columns

Index(['Unnamed: 0', 'Flow ID', 'Src IP', 'Src Port', 'Dst IP', 'Dst Port',
       'Protocol', 'Timestamp', 'Flow Duration', 'Tot Fwd Pkts',
       'Tot Bwd Pkts', 'TotLen Fwd Pkts', 'TotLen Bwd Pkts', 'Fwd Pkt Len Max',
       'Fwd Pkt Len Min', 'Fwd Pkt Len Mean', 'Fwd Pkt Len Std',
       'Bwd Pkt Len Max', 'Bwd Pkt Len Min', 'Bwd Pkt Len Mean',
       'Bwd Pkt Len Std', 'Flow Byts/s', 'Flow Pkts/s', 'Flow IAT Mean',
       'Flow IAT Std', 'Flow IAT Max', 'Flow IAT Min', 'Fwd IAT Tot',
       'Fwd IAT Mean', 'Fwd IAT Std', 'Fwd IAT Max', 'Fwd IAT Min',
       'Bwd IAT Tot', 'Bwd IAT Mean', 'Bwd IAT Std', 'Bwd IAT Max',
       'Bwd IAT Min', 'Fwd PSH Flags', 'Bwd PSH Flags', 'Fwd URG Flags',
       'Bwd URG Flags', 'Fwd Header Len', 'Bwd Header Len', 'Fwd Pkts/s',
       'Bwd Pkts/s', 'Pkt Len Min', 'Pkt Len Max', 'Pkt Len Mean',
       'Pkt Len Std', 'Pkt Len Var', 'FIN Flag Cnt', 'SYN Flag Cnt',
       'RST Flag Cnt', 'PSH Flag Cnt', 'ACK Flag Cnt', 'URG Flag Cnt',
       'CWE Flag

In [4]:
df.info(verbose=True)

<class 'dask.dataframe.core.DataFrame'>
Int64Index: 7616509 entries, 0 to 47062
Data columns (total 85 columns):
 #   Column             Non-Null Count  Dtype
---  ------             --------------  -----
 0   Unnamed: 0         7616509 non-null      int64
 1   Flow ID            7616509 non-null      object
 2   Src IP             7616509 non-null      object
 3   Src Port           7616509 non-null      int64
 4   Dst IP             7616509 non-null      object
 5   Dst Port           7616509 non-null      int64
 6   Protocol           7616509 non-null      int64
 7   Timestamp          7616509 non-null      object
 8   Flow Duration      7616509 non-null      int64
 9   Tot Fwd Pkts       7616509 non-null      int64
10   Tot Bwd Pkts       7616509 non-null      int64
11   TotLen Fwd Pkts    7616509 non-null      float64
12   TotLen Bwd Pkts    7616509 non-null      float64
13   Fwd Pkt Len Max    7616509 non-null      float64
14   Fwd Pkt Len Min    7616509 non-null      float64
15 

In [18]:
df.describe().compute()

c:\users\basti\documents\coding\python-venv\data-exploration-project\lib\site-packages\dask\array\reductions.py:692: RuntimeWarning: invalid value encountered in subtract
  xs = [sum((A - u) ** i, dtype=dtype, **kwargs) for i in range(2, order + 1)]
c:\users\basti\documents\coding\python-venv\data-exploration-project\lib\site-packages\numpy\lib\function_base.py:3961: RuntimeWarning: invalid value encountered in subtract
  diff_b_a = subtract(b, a)


Unnamed: 0      Src Port      Dst Port      Protocol  Flow Duration  \
count  7.616509e+06  7.616509e+06  7.616509e+06  7.616509e+06   7.616509e+06   
mean   3.518141e+06  4.171610e+04  1.030737e+04  9.066993e+00   1.168721e+07   
std    2.325547e+06  2.326642e+04  2.013310e+04  5.087024e+00   3.066261e+07   
min    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  -1.000000e+00   
25%    1.983526e+06  4.945800e+04  8.000000e+01  6.000000e+00   2.869000e+03   
50%    3.955070e+06  5.212200e+04  4.430000e+02  6.000000e+00   3.666950e+04   
75%    5.935024e+06  5.750000e+04  4.976750e+04  1.700000e+01   4.286910e+06   
max    7.902474e+06  6.553500e+04  6.553500e+04  1.700000e+01   1.200000e+08   

       Tot Fwd Pkts  Tot Bwd Pkts  TotLen Fwd Pkts  TotLen Bwd Pkts  \
count  7.616509e+06  7.616509e+06     7.616509e+06     7.616509e+06   
mean   1.314120e+01  6.606282e+00     6.304985e+02     5.230745e+03   
std    1.009401e+03  3.239508e+02     3.153876e+04     6.971802e+05   
min    0.000000e+00  0.000000e+00     0.000000e+00     0.000000e+00   
25%    1.000000e+00  1.000000e+00     2.900000e+01     0.000000e+00   
50%    2.000000e+00  3.000000e+00     4.600000e+01     2.470000e+02   
75%    8.000000e+00  7.000000e+00     9.350000e+02     1.039000e+03   
max    2.594430e+05  2.919230e+05     8.302176e+06     6.554527e+08   

       Fwd Pkt Len Max  Fwd Pkt Len Min  Fwd Pkt Len Mean  Fwd Pkt Len Std  \
count     7.616509e+06     7.616509e+06      7.616509e+06     7.616509e+06   
mean      2.196190e+02     1.186438e+01      5.794405e+01     8.231711e+01   
std       3.273044e+02     2.411128e+01      7.212265e+01     1.384465e+02   
min       0.000000e+00     0.000000e+00      0.000000e+00     0.000000e+00   
25%       2.800000e+01     0.000000e+00      1.033333e+01     0.000000e+00   
50%       4.200000e+01     0.000000e+00      3.900000e+01     1.154701e+01   
75%       9.350000e+02     3.500000e+01      2.337500e+02     4.675000e+02   
max       2.336000e+04     1.472000e+03      4.660441e+03     7.137621e+03   

       Bwd Pkt Len Max  Bwd Pkt Len Min  Bwd Pkt Len Mean  Bwd Pkt Len Std  \
count     7.616509e+06     7.616509e+06      7.616509e+06     7.616509e+06   
mean      3.571852e+02     2.843344e+01      1.150348e+02     1.310117e+02   
std       5.177774e+02     5.191029e+01      1.684583e+02     2.062547e+02   
min       0.000000e+00     0.000000e+00      0.000000e+00     0.000000e+00   
25%       0.000000e+00     0.000000e+00      0.000000e+00     0.000000e+00   
50%       2.470000e+02     0.000000e+00      7.600000e+01     1.220662e+02   
75%       7.590000e+02     6.200000e+01      1.777778e+02     2.440000e+02   
max       6.516000e+04     3.684000e+03      3.037508e+04     2.244841e+04   

        Flow Byts/s   Flow Pkts/s  Flow IAT Mean  Flow IAT Std  Flow IAT Max  \
count  7.586821e+06  7.616509e+06   7.616509e+06  7.616509e+06  7.616509e+06   
mean            inf           inf   2.225967e+06  1.134660e+06  5.318438e+06   
std             inf           inf   1.044042e+07  3.904521e+06  1.529950e+07   
min    0.000000e+00 -2.000000e+06  -1.000000e+00  0.000000e+00 -1.000000e+00   
25%    1.290649e+01  5.973010e+00   2.215000e+03  0.000000e+00  2.756500e+03   
50%    1.370277e+03  1.368691e+02   2.579550e+04  3.150545e+02  3.595350e+04   
75%    1.121417e+05  6.688963e+03   5.972362e+05  1.455863e+06  4.170336e+06   
max             inf           inf   1.199993e+08  8.480808e+07  1.199997e+08   

       Flow IAT Min   Fwd IAT Tot  Fwd IAT Mean   Fwd IAT Std   Fwd IAT Max  \
count  7.616509e+06  7.616509e+06  7.616509e+06  7.616509e+06  7.616509e+06   
mean   1.728327e+06  1.107273e+07  2.488425e+06  1.164599e+06  4.816895e+06   
std    1.034646e+07  3.048820e+07  1.061671e+07  4.396200e+06  1.493230e+07   
min   -1.300000e+01  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
25%    5.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
50%    5.200000e+01  1.399000e+03  5.013333e+02

In [ ]:
unique_list = []
for a in df.columns:
    unique_list.append(df[a].nunique().compute())
    print(a)

In [ ]:
for i in range(len(unique_list)):
    print(df.columns[i], "\t", unique_list[i])

### Unique Values
Unnamed: 0 	 6553320
Flow ID 	 4715999

Src IP 	 36760

Src Port 	 64849

Dst IP 	 34321

Dst Port 	 59091

Protocol 	 3

Timestamp 	 77598

Flow Duration 	 2722954

Tot Fwd Pkts 	 2501

Tot Bwd Pkts 	 2854

TotLen Fwd Pkts 	 14820

TotLen Bwd Pkts 	 60048

Fwd Pkt Len Max 	 3431

Fwd Pkt Len Min 	 479

Fwd Pkt Len Mean 	 98698

Fwd Pkt Len Std 	 210758

Bwd Pkt Len Max 	 3787

Bwd Pkt Len Min 	 444

Bwd Pkt Len Mean 	 165786

Bwd Pkt Len Std 	 243233

Flow Byts/s 	 3569113

Flow Pkts/s 	 3126852

Flow IAT Mean 	 3092172

Flow IAT Std 	 3619380

Flow IAT Max 	 1783978

Flow IAT Min 	 344330

Fwd IAT Tot 	 2260236

Fwd IAT Mean 	 2531071

Fwd IAT Std 	 2801288

Fwd IAT Max 	 1526963

Fwd IAT Min 	 443349

Bwd IAT Tot 	 2224031

Bwd IAT Mean 	 2444275

Bwd IAT Std 	 2605434

Bwd IAT Max 	 1426977

Bwd IAT Min 	 374503

Fwd PSH Flags 	 2

Bwd PSH Flags 	 2

Fwd URG Flags 	 1

Bwd URG Flags 	 1

Fwd Header Len 	 4225

Bwd Header Len 	 5132

Fwd Pkts/s 	 3013865

Bwd Pkts/s 	 2722380

Pkt Len Min 	 175

Pkt Len Max 	 4247

Pkt Len Mean 	 261546

Pkt Len Std 	 372908

Pkt Len Var 	 379741

FIN Flag Cnt 	 2

SYN Flag Cnt 	 2

RST Flag Cnt 	 2

PSH Flag Cnt 	 2

ACK Flag Cnt 	 2

URG Flag Cnt 	 2

CWE Flag Count 	 2

ECE Flag Cnt 	 2

Down/Up Ratio 	 213

Pkt Size Avg 	 204655

Fwd Seg Size Avg 	 71745

Bwd Seg Size Avg 	 137298

Fwd Byts/b Avg 	 1

Fwd Pkts/b Avg 	 1

Fwd Blk Rate Avg 	 1

Bwd Byts/b Avg 	 1

Bwd Pkts/b Avg 	 1

Bwd Blk Rate Avg 	 1

Subflow Fwd Pkts 	 2501

Subflow Fwd Byts 	 14820

Subflow Bwd Pkts 	 2854

Subflow Bwd Byts 	 60048

Init Fwd Win Byts 	 9520

Init Bwd Win Byts 	 13109

Fwd Act Data Pkts 	 749

Fwd Seg Size Min 	 10

Active Mean 	 704693

Active Std 	 542822

Active Max 	 580373

Active Min 	 287446

Idle Mean 	 863197

Idle Std 	 564679

Idle Max 	 599357

Idle Min 	 678358

Label 	 2

In [ ]:
# Anzahl der Nullwerte in der Spalte "Flow Byts/s".
df["Flow Byts/s"].isnull().sum().compute()

In [19]:
# Anzahl der inf-werte bestimmen pro Label
df[ (df["Flow Byts/s"] == np.inf) & (df["Flow Pkts/s"] == np.inf)].groupby("Label").agg(["count"]).compute()

Unnamed: 0 Flow ID Src IP Src Port Dst IP Dst Port Protocol Timestamp  \
            count   count  count    count  count    count    count     count   
Label                                                                          
Benign      18067   18067  18067    18067  18067    18067    18067     18067   

       Flow Duration Tot Fwd Pkts Tot Bwd Pkts TotLen Fwd Pkts  \
               count        count        count           count   
Label                                                            
Benign         18067        18067        18067           18067   

       TotLen Bwd Pkts Fwd Pkt Len Max Fwd Pkt Len Min Fwd Pkt Len Mean  \
                 count           count           count            count   
Label                                                                     
Benign           18067           18067           18067            18067   

       Fwd Pkt Len Std Bwd Pkt Len Max Bwd Pkt Len Min Bwd Pkt Len Mean  \
                 count           count           count            count   
Label                                                                     
Benign           18067           18067           18067            18067   

       Bwd Pkt Len Std Flow Byts/s Flow Pkts/s Flow IAT Mean Flow IAT Std  \
                 count       count       count         count        count   
Label                                                                       
Benign           18067       18067       18067         18067        18067   

       Flow IAT Max Flow IAT Min Fwd IAT Tot Fwd IAT Mean Fwd IAT Std  \
              count        count       count        count       count   
Label                                                                   
Benign        18067        18067       18067        18067       18067   

       Fwd IAT Max Fwd IAT Min Bwd IAT Tot Bwd IAT Mean Bwd IAT Std  \
             count       count       count        count       count   
Label                                                                 
Benign       18067       18067       18067        18067       18067   

       Bwd IAT Max Bwd IAT Min Fwd PSH Flags Bwd PSH Flags Fwd URG Flags  \
             count       count         count         count         count   
Label                                                                      
Benign       18067       18067         18067         18067         18067   

       Bwd URG Flags Fwd Header Len Bwd Header Len Fwd Pkts/s Bwd Pkts/s  \
               count          count          count      count      count   
Label                                                                      
Benign         18067          18067          18067      18067      18067   

       Pkt Len Min Pkt Len Max Pkt Len Mean Pkt Len Std Pkt Len Var  \
             count       count        count       count       count   
Label                                                                 
Benign       18067       18067        18067       18067       18067   

       FIN Flag Cnt SYN Flag Cnt RST Flag Cnt PSH Flag Cnt ACK Flag Cnt  \
              count        count        count        count        count   
Label                                                                     
Benign        18067        18067        18067        18067        18067   

       URG Flag Cnt CWE Flag Count ECE Flag Cnt Down/Up Ratio Pkt Size Avg  \
              count          count        count         count        count   
Label                                                                        
Benign        18067          18067        18067         18067        18067   

       Fwd Seg Size Avg Bwd Seg Size Avg Fwd Byts/b Avg Fwd Pkts/b Avg  \
                  count            count          count          count   
Label                                                                    
Benign            18067            18067          18067          18067   

       Fwd Blk Rate Avg Bwd Byts/b Avg Bwd Pkts/b Avg Bwd Blk Rate Avg  \
                  count          count          coun

# Erste Erkenntnisse

## Unnötige Daten
Um nähere Infos über die Spalten zu erhalten, haben wir uns die Details und Anzahl der Unqiue-Werte pro Spalte ermittelt.
Folgende Spalten haben lediglich einen Uniquen-Wert:
- Fwd URG Flags 
- Bwd URG Flags 
- Fwd Byts/b Avg
- Fwd Pkts/b Avg
- Bwd Byts/b Avg
- Bwd Pkts/b Avg
- Fwd Blk Rate Avg
- Bwd Blk Rate Avg

Da diese Attribute lediglich einen Wert annehmen können, sind diese nicht hilfreich bei der Unterscheidung zwischen gutartig & DDoS. Im folgenden werden diese Werte daher aus dem Datensatz entfernt --> 10% der ursprünglichen Spalten entfernt.

## Null-Values
- Flow Byts/s
Die Spalte Flow Byts/s hat als einzige Spalte 29688 Nullwerte (nicht definiert NaN), was wir durch .info erkannt haben. Es gibt zwei Möglichkeiten, damit umzugehen.
- Option 1: Droppen
  - Ein Droppen wäre sinnvoll, da wir dadurch keine falschen Annahmen über die Daten treffen. Zudem machen diese Daten gerade erst einmal 0,4 % der Daten aus. 
- Option 2: Werte ersetzen 
  - Das Ersetzen der Werte würde die Timeline (timestamps) nicht künstlich verändern (evtl. problematisch bei manchen ML-Modellen).

Vorest entscheiden wir uns für Option 1.

## Infinity-Values
Bei der Untersuchung ist uns aufgefallen, dass Flow Byts/s und Flow Pkts/s in 0,2% der Daten Infinity-Werte aufweisen. Bei diesen Daten handelt es sich gutartige (Benign) Anfragen handelt. Deshalb haben wir uns dazu entschieden, diese Daten zu entfernen.

In [5]:
df_inf = df[~da.isinf(df["Flow Byts/s"])]
df_nonull = df_inf.dropna(subset=["Flow Byts/s"])
df_clean = df_nonull.drop(["Fwd URG Flags", "Bwd URG Flags", "Fwd Byts/b Avg", "Fwd Pkts/b Avg", "Bwd Byts/b Avg", "Bwd Pkts/b Avg", "Fwd Blk Rate Avg", "Bwd Blk Rate Avg"], axis=1)
df_clean.to_csv('cleaned_ds.csv', single_file = True, index = False)

['D:/applications/ddos-dompteure/cleaned_ds.csv']